homeassistant/helpers/config_validation.py

```diff
@@ -73,3 +106,93 @@ def time_zone(value):
     raise vol.Invalid(
         'Invalid time zone passed in. Valid options can be found here: '
         'http://en.wikipedia.org/wiki/List_of_tz_database_time_zones')
+
+
+# Validator helpers
+
+# pylint: disable=too-few-public-methods
+
+class DictValidator(object):
+    """Validate keys and values in a dictionary."""
+
+    def __init__(self, value_validator=None, key_validator=None):
+        """Initialize the dict validator."""
+        if value_validator is not None:
+            value_validator = vol.Schema(value_validator)
+
+        self.value_validator = value_validator
+
+        if key_validator is not None:
+            key_validator = vol.Schema(key_validator)
+
+        self.key_validator = key_validator
+
+    def __call__(self, obj):
+        """Validate the dict."""
+        if not isinstance(obj, dict):
+            raise vol.Invalid('Expected dictionary.')
+
+        errors = []
+
+        # So we keep it an OrderedDict if it is one
+        result = obj.__class__()
+
+        for key, value in obj.items():
+            if self.key_validator is not None:
+                try:
+                    key = self.key_validator(key)
+                except vol.Invalid as ex:
+                    errors.append('key {} is invalid ({})'.format(key, ex))
+
+            if self.value_validator is not None:
+                try:
+                    value = self.value_validator(value)
+                except vol.Invalid as ex:
+                    errors.append(
+                        'key {} contains invalid value ({})'.format(key, ex))
+
+            if not errors:
+                result[key] = value
+
+        if errors:
+            raise vol.Invalid(
+                'invalid dictionary: {}'.format(', '.join(errors)))
+
+        return result
+
+
+# Adapted from:
+# https://github.com/alecthomas/voluptuous/issues/115#issuecomment-144464666
+def has_at_least_one_key(keys):
```

Comment:
Nice, I think this is exactly what I was struggling with with the MQTT service payload/payload_template argument, I couldn't figure out how to make it both exclusive and required.